Implements efx api node, working on Kovan + mocked HTTP API

hems opened the pull request:
````js
const efx = await EFX()
````

 - Client api
    - I refactored all the code samples into functions and puzzled them together
http methods stays on the main namespace:
      - efx.getOrder(orderId)
      - efx.submitOrder(token,amount,duration)
      - ...

    - Contract related methods are under '.contract' namesapce, for instance:
        - efx.contract.lock(token, amount, duration)
        - efx.contract.unlock(token,amount)
        - ...

    - Sign functions under .sign
        - efx.sign(anything)
        - efx.sign.order(symbol, amount, price)
        - ...

    - ...

 - Tests
    - Thanks to [ganache](https://github.com/trufflesuite/ganache-cli) and [nock](https://github.com/nock/nock) we should be able to test the client, independently, offline and very fast !

    - A few tests are already done and i believe nearly all tests are scaffolded just waiting for final details, [have a look a the build logs to get and idea!](https://travis-ci.org/hems/efx-api-node/jobs/400148651)

    - Currently the code also works on the browser, you can toy around on your browser console with `npm run build:web:run`

    - You can run node.js tests with: `npm run test` and browser tests with `npm run test:web`
      - Unfortunately `nock` breaks the browser tests, so we need to replace nock by [xmr-mock](https://www.npmjs.com/package/xhr-mock) or mock ( using express for instance ) the ethfinex service

 - Web build
    - Thanks to [browserify](http://browserify.org/), the standalone web build is working so the library could be imported via [RawGit](https://rawgit.com/) or similar

Next steps:

 - FIX `efx.contract.lock` and `efx.contract.unlock` tests, there is something funny happening on the code that calls the contract.

 - FIX / REVIEW ecrecover from signature test

 - Populate [list of with all currencies](https://github.com/hems/efx-api-node/blob/master/src/currencies.js)? 

 - Test client against the real http api

 - Review [method implementations](https://github.com/hems/efx-api-node/tree/master/src/api)

 - Add more examples and documentation regarding the client on the README.md file

 - Deal better with errors / selecting ( default ) eth account

 - Check if user has tokens locked before submiting an order

hems commented on src/api/account/select.js:
```diff
@@ -0,0 +1,25 @@
+ **/
+module.exports = async (client, id) => {
+  // check for ethereum accounts and select a default one
+  const accounts = await client.web3.eth.getAccounts()
```
@plutoegg Currently i'm automatically selecting accounts[0] but i have the impression that might to be ideal? Perhaps we should look for https://github.com/ethereum/wiki/wiki/JavaScript-API#web3ethdefaultaccount and if it doesn't exist we try accounts[0] ?

hems commented on src/api/cancel_order.js:
```diff
@@ -0,0 +1,12 @@
+  const data = {
+    orderId: id,
+    signature: 'web3signature'
+  }
```
@plutoegg / @prdn do we have more specs for this? 

Or do we simply sign `{ orderId }` ?

plutoegg replied:
Just orderId should be fine

hems commented on src/api/contract/abi.js:
```diff
@@ -0,0 +1,64 @@
+ * The ABI for the wrapper token contract
+ *
+ */
+module.exports = [
```
@plutoegg can you gist me the latest abi?

hems commented on src/api/contract/errors.js:
```diff
@@ -0,0 +1,46 @@
+module.exports = {
+
+  ERR_MAKERTOKEN_ADDRESS_INVALID: `
```
@prdn / @plutoegg imported the error codes and descriptions here I thought it could be useful to yield user-friendly messages. thoughts?

plutoegg replied:
I think definitely a good approach to transform the error codes into user friendly ones!

hems commented on src/api/contract/lock.js:
```diff
@@ -0,0 +1,33 @@
+  ]
+
+  if (action === 'lock') {
+    if (token === 'ETH') {
```
@plutoegg I believe `token == 'ETH'` isn't the right thing to do here right?

hems commented on src/api/eth/call.js:
```diff
@@ -0,0 +1,13 @@
+  const { web3 } = efx
+
+  return new Promise((resolve, reject) => {
+    const contract = new web3.eth.Contract(wrapperAbi).at(address)
```
@plutoegg this is where the [tests are failing](https://travis-ci.org/hems/efx-api-node/jobs/400148651#L654) should be an easy fix.... ( i guess )

hems commented on src/api/eth/get_network.js:
```diff
@@ -0,0 +1,26 @@
+const P = require('bluebird')
+
+module.exports = async (efx) => {
+  const get = P.promisify(efx.web3.version.getNetwork)
```
@prdn / @plutoegg the idea here is to export a util function so we could for instance load a different default configuration based on the eth network we are connected to.

hems commented on src/api/register_orders.js:
```diff
@@ -0,0 +1,15 @@
+const { post } = require('request-promise')
+
+module.exports = (client, id) => {
+  const url = client.config.api + '/registerOrders'
```
@plutoegg / @prdn do we have more specs on this?

hems commented on src/api/release_tokens.js:
```diff
@@ -0,0 +1,13 @@
+const { post } = require('request-promise')
+
+module.exports = (client, id) => {
+  const url = client.config.api + '/releaseTokens'
```
@prdn / @plutoegg will the signature be `releaseTokens(token, amount)` ?

hems commented on src/api/sign/cancel_order.js:
```diff
@@ -0,0 +1,9 @@
+
+  const toSign = utils.bufferToHex(orderId).slice(2)
+
+  return efx.sign(toSign)
```
@plutoegg i still need to sort this .sign jazz, will look into it soon

hems commented on src/api/sign/request.js:
```diff
@@ -0,0 +1,8 @@
+    // headers: new window.Headers(),
+    // mode: 'cors',
+    // cache: 'default'
+    json: JSON.stringify({contents})
```
@plutoegg / @prdn do we really mean `JSON.stringify( { contents } )` and not `JSON.strifiy(contents)` ?

hems commented on src/api/sign/sign.js:
```diff
@@ -0,0 +1,22 @@
+module.exports = (client, toSign) => {
+  const { config, web3 } = client
+
+  const sign = P.promisify(web3.eth.sign)
```
this guy is to blame on the weird signatures 🖌

hems commented on src/api/submit_order.js:
```diff
@@ -0,0 +1,32 @@
+  // REVIEW: shall we use signed.ecSignature ?
+  // or refactor it to just signed.signature ?
+  // or even just return the signature? ( this would be problematic for tests )
+  const meta = { protocol: '0x', object: signed.signature }
```
@prdn / @plutoegg I guess it's on the `object:` field that we will attach the signature?

plutoegg replied:
The object needs to contain the full order as well as the signature

i.e. `object: signedOrder`

signOrder contains all params, as well as an additional field ecSignature
```
const signedOrder = {
    expirationUnixTimestampSec: ...,
    feeRecipient: ...,
    maker: ...,
    ecSignature: {
       v: ...,
       r: ...,
       s: ...,
    },
}

```

hems replied:
okie

hems commented:
I figured out I was doing the calls to the contracts incorrectly, now they are correct or as close to correct I could get on my tests, as a consequence I refactored the contract calls/send lock/unlock code.

Need to check with @plutoegg the address of the WrapperLockEth on the Kovan network.

Next step is to figure out the tests for the signatures in order to make sure they work as expected and then all the important bits will be cooked and working.

🍡

plutoegg commented on src/api/contract/unlock.js:
```diff
@@ -0,0 +1,24 @@
+  const s = signature.s || 0
+  const v = signature.v || 0
+
+  const args = [v, r, s, value, unlockTill]
```
The signature required to unlock with the withdraw function should not be signed by the user's own key.

Instead it needs to be signed by Ethfinex. To do this, the user needs to request a signature from the releaseTokens endpoint

The signature will contain: keccak256(userAddress, wrapperAddress, signatureValidUntilBlock)

hems replied:
@plutoegg do i need to send a "unlockUntil" to the http endpoint /releaseTokens ?

hems replied:
I'm not exactly sure what i should use as  "signatureValidUntilBlock" on my test file

plutoegg commented on src/api/eth/send.js:
```diff
@@ -0,0 +1,13 @@
+
+  // REVIEW: Shall we use estimateGas here? maybe show the user what the gas
+  // will be?
+  // see: https://web3js.readthedocs.io/en/1.0/web3-eth-contract.html#methods-mymethod-estimategas
```
@hems We can estimate gas but the gas of most calls (deposit/withdraw) to lock and unlock should be fixed so we might be happy to set these as fixed to avoid using estimateGas.

hems replied:
@plutoegg i can see if we don't provide gas/gasPrice it does call the estimateGas automatically. 
I guess fixing those values would simply save us some time on each send request.

plutoegg commented on src/api/release_tokens.js:
```diff
@@ -0,0 +1,13 @@
+  const data = {
+    address: '0x01...',
+    tokenAddress: '0x02...',
+    unlockUntil: 123456
```
@hems I don't think we want unlockUntil to be specifiable - we will just give the user 10 minutes or so from whenever they submit the request. (Ignore the readme which is outdated in this case)

hems replied:
@plutoegg okay then, I'm removing the unlock until value from the payload

plutoegg commented on src/api/sign/lock.js:
```diff
@@ -0,0 +1,28 @@
+  }
+
+  const url = config.api // do we add anything to this path?
+
```
@hems think this is the wrong way around for locking (and payload and request is just for unlocking)?

hems replied:
@plutoegg indeed for locking the user will interact directly with web3, this is already updated on the latest version which i'm about to push.

plutoegg commented on src/api/submit_order.js:
```diff
@@ -0,0 +1,32 @@
+    throw new Error('symbol, amount and price are required')
+  }
+
+  gid = gid || 1 // REVIEW: shall we make it unique as well?
```
@prdn ?

plutoegg commented on src/config.js:
```diff
@@ -0,0 +1,19 @@
+
+  exchangeContractAddress: '0x0000000000000000000000000000000000000000',
+
+  zxAddress: '0x',
```
@hems is this required?

hems replied:
@plutoegg zxAddress is not necessary, i removed now. exchangeContract is used by sign/order.js

hems commented:
Since the ETH `withdrawal(amount)` only takes `amount` the `.unlock` method already works.

For tokens i believe we will neese the `releaseTokens` method from the HTTP API working so we can get the signature in order to withdarawal tokens?

Also for `submitOrder` i guess it won't work before we have the HTTP API in place?

hems commented:
I have now reviewed the `efx-pub-api.js` file and refactored the HTTP calls to reflect the backend.

registerOrderList, submitOrder, releaseTokens and cancelOrder now post exactly the same payload as `efx-pub-api` expects and sign the requests accordingly.

I have successfully validated the signatures using ecRecover, given `efx-pub-api` parameters haven't changed i expect those methods to be working.

From the README @plutoegg gave me i noticed the pub-api still don't implement:
 - `getOrder(orderId)`
 - `getOrders()`
 - `getPendingOrders()`

Once those are in place i can  mock some responses for the tests, but it also should be working.

@plutoegg [please let me know if the algorithm i implemented for signOrder is the correct one](https://github.com/hems/efx-api-node/blob/master/src/api/sign/cancel_order.js).

You can see [on this test](https://github.com/hems/efx-api-node/blob/0b59a580c155228d8a6d67327852f6906cba0f22/test/http-api.js#L22-L37) the ecRecover is now working as expected, so changing the algo should be the easy part.

Would be nice to get a list of `tokenAddress` and `wrapperAddress` for all the tokens on the Kovan ( and maybe on the mainet too? )

hems commented:
In order to have ecRecover native in node.js ( don't rely on an RPC call ) I end up having to change how I sign a couple of payloads.

Namely
 - `ZeroEx's orderHash`
![image](https://user-images.githubusercontent.com/27327/42553751-fda1b184-84d9-11e8-8ec2-b7af2f1d7b51.png)

Reason: signing an hex strings was yielding an error when ecRecover-ing

 - `registerOrderList's request`
![image](https://user-images.githubusercontent.com/27327/42553782-2a34630e-84da-11e8-944b-e1cbe4e746b7.png)

Reason: A object was being signed here with web3.eth.sign, but turns out when ecRecovering using "JSON.stringify" as serialization message was yielding an error when ecRecover-ing as well

The js-server would have to adapt or we would have to fix this ecRecovering routine to work with objects and hex strings. 

I noticed @plutoegg was removing the 0x before signing bufferToHex(OrderId), so i think now the implementations match.
